consensus/XDPoS/engines/engine_v2/engine.go

```diff
@@ -1133,22 +1165,38 @@ func (x *XDPoS_v2) getEpochSwitchInfo(chain consensus.ChainReader, header *types
 	}
 	if isEpochSwitch {
 		log.Debug("[getEpochSwitchInfo] header is epoch switch", "hash", hash.Hex(), "number", h.Number.Uint64())
-		masternodes := x.GetMasternodesFromEpochSwitchHeader(h)
-		// create the epoch switch info and cache it
-		var decodedExtraField utils.ExtraFields_v2
-		err = utils.DecodeBytesExtraFields(h.Extra, &decodedExtraField)
-		if err != nil {
-			return nil, err
-		}
-		epochSwitchInfo := &utils.EpochSwitchInfo{
-			Masternodes: masternodes,
-			EpochSwitchBlockInfo: &utils.BlockInfo{
-				Hash:   hash,
-				Number: h.Number,
-				Round:  decodedExtraField.Round,
-			},
-			EpochSwitchParentBlockInfo: decodedExtraField.QuorumCert.ProposedBlockInfo,
+		var epochSwitchInfo *utils.EpochSwitchInfo
+		// Special case, in case of last v1 block, we manually build the epoch switch info
```

Comment:
This `if` should use recently merged XIN-127 to get masternodes, now that XIN-127 finished init snapshot

Reply:
Discussed with @wjrjerome that snapshot is only recorded the initial master nodes without penalty and comeback nodes. Also snapshot is local generated, but this function is getting call when block comes from external.